File: viewer/packages/pointclouds/wasm/src/point_octree.rs
use crate::Vec3WithIndex;

use std::vec::Vec;

mod octree_node;
use octree_node::*;

use crate::linalg::{Vec3, BoundingBox};

pub struct PointOctree {
    root: OctreeNode
}

impl PointOctree {
    pub fn new(points: Vec<Vec3WithIndex>, bounding_box: BoundingBox) -> PointOctree {
        PointOctree {
            root: OctreeNode::new(points, bounding_box)
        }
    }

    pub fn get_points_in_box(&self, bounding_box: &BoundingBox) -> Vec<Vec3WithIndex> {
        let mut v: Vec<Vec3WithIndex> = Vec::new();

        self.root.get_points_in_box(bounding_box, &mut v);

        v
    }
}

File: viewer/packages/pointclouds/src/pointOctree/PointOctree.ts
/*!
 * Copyright 2022 Cognite AS
 */

import { Vec3WithIndex } from '../styling/shapes/linalg';

import * as THREE from 'three';
import { OctreeNode } from './OctreeNode';

export class PointOctree {
  private readonly _root: OctreeNode;

  constructor(points: Float64Array, boundingBox: THREE.Box3) {
    const pointArray = new Array<Vec3WithIndex>(points.length / 3);
    for (let i = 0; i < points.length / 3; i++) {
      pointArray[i] = [points[3 * i + 0], points[3 * i + 1], points[3 * i + 2], i];
    }

    this._root = new OctreeNode(pointArray, boundingBox);
  }

  getPointsInBox(box: THREE.Box3): Vec3WithIndex[] {
    const list: Vec3WithIndex[] = [];

    this._root.getPointsInBox(box, list);

    return list;
  }

  get root(): OctreeNode {
    return this._root;
  }
}
